File: scala/92.scala
object Solution {
    def reverseBetween(head: ListNode, left: Int, right: Int): ListNode = {
        def reverse(head: ListNode, k: Int): ListNode = {
          if (head == null) return head
          if (k < left || k >= right) {
            head.next = reverse(head.next, k+1)
            head
          }
          else {
            val tail = reverse(head.next, k+1)
            val s = head.next.next
            head.next.next = head
            head.next = s
            tail
          }
        }
        reverse(head, 1)
    }
}

File: leetcodeJava/92.java
package leetcodeJava;

class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode tail = null, tail2 = null;
        var dummyHead = new ListNode(0, head);
        var ptr = dummyHead;
        var index = 0;
        while (ptr != null) {
            var next = ptr.next;
            if (index == left-1) {
                tail = ptr;
                tail.next = null;
            }
            if (index >= left && index <= right) {
                if (tail.next == null) tail2 = ptr;
                ptr.next = tail.next;
                tail.next = ptr;
            }
            if (index == right) {
                tail2.next = next;
                break;
            }
            index++;
            ptr = next;
        }
        return dummyHead.next;
    }
}
